### ISSUE TITLE
Add stubs for cython to typeshed

### ISSUE THREAD

**Issue by OswinC** [cython/cython#1826]
Hi,
We are using Cython in some parts of our projects. And we also use the type annotation feature of Python for our Python projects. To unblock type checking over the boundary between Python code and Cython code, we created pyi files for our Cython code so that the type checkers like mypy could find the names defined in the Cython code. We also created pyi files for cython.py and would like to contribute it back to the typeshed upstream. Are you guys good with us adding the stubs for Cython to typeshed? https://github.com/python/typeshed/pull/1554

**Comment by scoder** [comment]
Thanks for doing that!

As far as I understand typeshed, they are just a repository for keeping third party stubs. I guess the reason why we would want to store those stubs externally, rather than shipping them with Cython, is that people who use your code uncompiled then wouldn't need to install Cython as a dependency just for the annotations. That would mostly solve the problem of packages trying to concurrently install a version of cython.py (a.k.a. Cython/Shadow.py) in order to use Cython type declarations in their Python code.

One drawback is that we'd have to reach out to typeshed in order to update the list of types (which tends to be rare and certainly hasn't happened in a while).

I'm also not sure about the impact on issue #1672. It might actually just work ;) - would you have an example of actual code that uses these annotations? Are you writing your code in Python syntax to allow optional compilation with Cython?

**Comment by ambv** [comment]
@scoder Facebook has an internal tool that we want to open source (the cython stubs are the first part of this effort). The tool is generating Python .pyi files from Cython sources. This enables type checking of code that uses .so files produced by Cython without mypy having to understand Cython syntax. This doesn't solve type checking of Cython files themselves, that will come later.

We're not really using the pure Python syntax anywhere outside of the generated .pyi files, maybe we're missing some parts of the Cython API in the proposed pull request to typeshed.

Anyway, I'll keep you and @robertwb in the loop for open-sourcing of this tool, it's off topic for this issue.

**Comment by OswinC** [comment]
In fact, the idea of our work is similar to shalabhc's suggestion in #1672. Here's an example-
Let's say that in foo.pyx, we have
`cpdef int str2int(str s):
      return int(s)`
then in the stub file foo.pyi for foo.pyx, we'll have
`def str2int(s: cython.str) -> cython.int: ...`

**Comment by scoder** [comment]
Just for completeness, here's what Cython currently accepts as annotation typing:
https://github.com/cython/cython/blob/0.26/tests/run/annotation_typing.pyx
Now that there's an actual way to represent its types in a PEP 484 conforming way, I think we should strip that down to what `typing` actually allows.

I think it's a good idea to generate .pyi files to avoid source code parsing. Does that tool generate them from the available runtime signature information or from source code? (If the latter, it feels like Cython should be doing that itself.)

**Comment by ambv** [comment]
> Does that tool generate them from the available runtime signature information or from source code? (If the latter, it feels like Cython should be doing that itself.)

From source code. We thought about integrating that with Cython itself but the current parsing model was hard for a new comer to grasp.

We can discuss that project in a separate issue, for now we need an "A-OK" from you for including your stubs in typeshed so we can merge that pull request :)

**Comment by scoder** [comment]
> We thought about integrating that with Cython itself but the current parsing model was hard for a new comer to grasp.

You could have asked. ;)
It wouldn't integrate with the parser directly, but treated as a fourth output format of the compiler (c/c++, HTML, API pxd/h, plus then API pyi). And there should probably also be an option to automatically copy the argument types into the runtime accessible `__annotations__`. But yes, let's keep that out of this ticket.

**Comment by scoder** [comment]
> including your stubs in typeshed

As I said, I'm very happy that this was done. I'm just trying to understand the implications, especially for maintenance. All I could find on the github page regarding an ownership policy was "if something disagrees with the implementation, fix it and file a pull request".

The type system of Cython is pretty stable, but there have been additions every once in a while. The description on [stub versioning](https://github.com/python/typeshed/blob/master/CONTRIBUTING.md#stub-versioning) says "in its current implementation, typeshed cannot contain stubs for multiple versions of the same third-party library".

Additions to the Cython type system tend to be backwards compatible (especially at the high level of a Python representation), which means that we'd normally just have to add them to the stubs and wait for type checkers to update their typeshed version. Sounds acceptable.

I'm fine with adding the stubs to typeshed.

**Comment by xhochy** [comment]
> @scoder Facebook has an internal tool that we want to open source (the cython stubs are the first part of this effort). The tool is generating Python .pyi files from Cython sources. This enables type checking of code that uses .so files produced by Cython without mypy having to understand Cython syntax. This doesn't solve type checking of Cython files themselves, that will come later.

@ambv Has this tool ever been released or is there nowadays still a chance this will ever be released?

**Comment by gramster** [comment]
From what I heard that tool didn't work well and was abandoned.
